File: demogithub/scripts/update_readme.py
import re
import random
from datetime import datetime, timezone, timedelta

QUOTES = [
    ("“Any fool can write code that a computer can understand. Good programmers write code that humans can understand.”", "Martin Fowler"),
    ("“First, solve the problem. Then, write the code.”", "John Johnson"),
    ("“Experience is the name everyone gives to their mistakes.”", "Oscar Wilde"),
    ("“Code is like humor. When you have to explain it, it’s bad.”", "Cory House"),
    ("“Simplicity is prerequisite for reliability.”", "Edsger W. Dijkstra"),
    ("“Make it work, make it right, make it fast.”", "Kent Beck"),
    ("“The best error message is the one that never shows up.”", "Thomas Fuchs"),
    ("“Architecture is about the important stuff. Whatever that is.”", "Ralph Johnson"),
]

def get_vietnam_time():
    # GMT+7
    tz_vietnam = timezone(timedelta(hours=7))
    return datetime.now(tz_vietnam)

def get_greeting(hour: int) -> str:
    if 5 <= hour < 12:
        return "🌅 **Chào buổi sáng!** Chúc bạn một ngày mới tràn đầy năng lượng & sáng tạo!"
    elif 12 <= hour < 18:
        return "☀️ **Chào buổi chiều!** Chúc bạn làm việc hiệu quả & giải quyết bài toán mượt mà!"
    elif 18 <= hour < 23:
        return "🌆 **Chào buổi tối!** Chúc bạn có thời gian thư giãn tuyệt vời!"
    else:
        return "🌙 **Cú đêm coding!** Đừng quên giữ gìn sức khỏe và nghỉ ngơi hợp lý nhé!"

def get_day_name(weekday: int) -> str:
    days = ["Thứ Hai", "Thứ Ba", "Thứ Tư", "Thứ Năm", "Thứ Sáu", "Thứ Bảy", "Chủ Nhật"]
    return days[weekday]

def generate_realtime_content() -> str:
    now = get_vietnam_time()
    greeting = get_greeting(now.hour)
    day_str = get_day_name(now.weekday())
    time_str = now.strftime("%H:%M:%S")
    date_str = now.strftime("%d/%m/%Y")
    
    quote_text, quote_author = random.choice(QUOTES)
    
    # URL encoded time badge
    time_badge_label = f"{time_str} (GMT+7)".replace("-", "--").replace(" ", "%20")
    date_badge_label = f"{day_str}, {date_str}".replace("-", "--").replace(" ", "%20")

    content = f"""<!-- REALTIME_START -->
<div align="center">
  <table>
    <tr>
      <td align="center">
        <img src="https://img.shields.io/badge/🟢_LIVE_STATUS-ONLINE-success?style=flat-square&color=10B981" alt="Live Status" />
        &nbsp;
        <img src="https://img.shields.io/badge/⏰_GIỜ_VIỆT_NAM-{time_badge_label}-6366F1?style=flat-square" alt="Vietnam Time" />
        &nbsp;
        <img src="https://img.shields.io/badge/📅_NGÀY-{date_badge_label}-3B82F6?style=flat-square" alt="Date" />
        <br/><br/>
        <p><i>{greeting}</i></p>
        <p>💡 {quote_text} — <b>{quote_author}</b></p>
        <sub style="color: #94A3B8;">⚡ Tự động cập nhật qua GitHub Actions • Cập nhật lần cuối: <b>{date_str} {time_str} (GMT+7)</b></sub>
      </td>
    </tr>
  </table>
</div>
<!-- REALTIME_END -->"""
    return content

def update_readme():
    readme_path = "README.md"
    with open(readme_path, "r", encoding="utf-8") as f:
        content = f.read()

    new_realtime_block = generate_realtime_content()
    pattern = r"<!-- REALTIME_START -->.*?<!-- REALTIME_END -->"

    if re.search(pattern, content, flags=re.DOTALL):
        updated_content = re.sub(pattern, new_realtime_block, content, flags=re.DOTALL)
    else:
        # If not present, insert below the quick social links or header
        if "</div>\n\n---" in content:
            updated_content = content.replace("</div>\n\n---", f"</div>\n\n{new_realtime_block}\n\n---", 1)
        else:
            updated_content = new_realtime_block + "\n\n" + content

    with open(readme_path, "w", encoding="utf-8") as f:
        f.write(updated_content)

    print(f"Successfully updated README.md with real-time stats at {get_vietnam_time().strftime('%Y-%m-%d %H:%M:%S (GMT+7)')}")

if __name__ == "__main__":
    update_readme()

File: demogithub/scripts/update_readme.js
const fs = require('fs');
const path = require('path');

const QUOTES = [
  { text: "Any fool can write code that a computer can understand. Good programmers write code that humans can understand.", author: "Martin Fowler" },
  { text: "First, solve the problem. Then, write the code.", author: "John Johnson" },
  { text: "Experience is the name everyone gives to their mistakes.", author: "Oscar Wilde" },
  { text: "Code is like humor. When you have to explain it, it’s bad.", author: "Cory House" },
  { text: "Simplicity is prerequisite for reliability.", author: "Edsger W. Dijkstra" },
  { text: "Make it work, make it right, make it fast.", author: "Kent Beck" },
  { text: "The best error message is the one that never shows up.", author: "Thomas Fuchs" },
  { text: "Architecture is about the important stuff. Whatever that is.", author: "Ralph Johnson" }
];

function getVietnamDate() {
  // Current UTC time + 7 hours
  const now = new Date();
  const utcTime = now.getTime() + (now.getTimezoneOffset() * 60000);
  return new Date(utcTime + (7 * 3600000));
}

function getGreeting(hour) {
  if (hour >= 5 && hour < 12) {
    return "🌅 **Chào buổi sáng!** Chúc bạn một ngày mới tràn đầy năng lượng & sáng tạo!";
  } else if (hour >= 12 && hour < 18) {
    return "☀️ **Chào buổi chiều!** Chúc bạn làm việc hiệu quả & giải quyết bài toán mượt mà!";
  } else if (hour >= 18 && hour < 23) {
    return "🌆 **Chào buổi tối!** Chúc bạn có thời gian thư giãn tuyệt vời!";
  } else {
    return "🌙 **Cú đêm coding!** Đừng quên giữ gìn sức khỏe và nghỉ ngơi hợp lý nhé!";
  }
}

function getDayName(dayIndex) {
  const days = ["Chủ Nhật", "Thứ Hai", "Thứ Ba", "Thứ Tư", "Thứ Năm", "Thứ Sáu", "Thứ Bảy"];
  return days[dayIndex];
}

function padZero(num) {
  return String(num).padStart(2, '0');
}

function generateRealtimeContent() {
  const vnTime = getVietnamDate();
  const hours = vnTime.getHours();
  const greeting = getGreeting(hours);
  const dayStr = getDayName(vnTime.getDay());
  
  const timeStr = `${padZero(hours)}:${padZero(vnTime.getMinutes())}:${padZero(vnTime.getSeconds())}`;
  const dateStr = `${padZero(vnTime.getDate())}/${padZero(vnTime.getMonth() + 1)}/${vnTime.getFullYear()}`;
  
  const randomQuote = QUOTES[Math.floor(Math.random() * QUOTES.length)];

  const timeBadge = encodeURIComponent(`${timeStr} (GMT+7)`).replace(/-/g, '--');
  const dateBadge = encodeURIComponent(`${dayStr}, ${dateStr}`).replace(/-/g, '--');

  return `<!-- REALTIME_START -->
<div align="center">
  <table>
    <tr>
      <td align="center">
        <img src="https://img.shields.io/badge/🟢_LIVE_STATUS-ONLINE-success?style=flat-square&color=10B981" alt="Live Status" />
        &nbsp;
        <img src="https://img.shields.io/badge/⏰_GIỜ_VIỆT_NAM-${timeBadge}-6366F1?style=flat-square" alt="Vietnam Time" />
        &nbsp;
        <img src="https://img.shields.io/badge/📅_NGÀY-${dateBadge}-3B82F6?style=flat-square" alt="Date" />
        <br/><br/>
        <p><i>${greeting}</i></p>
        <p>💡 “${randomQuote.text}” — <b>${randomQuote.author}</b></p>
        <sub style="color: #94A3B8;">⚡ Tự động cập nhật qua GitHub Actions • Cập nhật lần cuối: <b>${dateStr} ${timeStr} (GMT+7)</b></sub>
      </td>
    </tr>
  </table>
</div>
<!-- REALTIME_END -->`;
}

function updateReadme() {
  const readmePath = path.join(__dirname, '..', 'README.md');
  let content = fs.readFileSync(readmePath, 'utf8');

  const newRealtimeBlock = generateRealtimeContent();
  const pattern = /<!-- REALTIME_START -->[\s\S]*?<!-- REALTIME_END -->/;

  if (pattern.test(content)) {
    content = content.replace(pattern, newRealtimeBlock);
  } else {
    // Insert right after the top header & social links block
    if (content.includes('</div>\n\n---')) {
      content = content.replace('</div>\n\n---', `</div>\n\n${newRealtimeBlock}\n\n---`);
    } else {
      content = `${newRealtimeBlock}\n\n${content}`;
    }
  }

  fs.writeFileSync(readmePath, content, 'utf8');
  console.log(`✅ Successfully updated README.md with real-time stats (GMT+7)!`);
}

updateReadme();
